File: ruby/space-age/space_age.rb
class SpaceAge
  AGE_CONVERSION = {
    mercury: 31557600 * 0.2408467,
    venus: 31557600 * 0.61519726,
    earth: 31557600,
    mars: 31557600 * 1.8808158,
    jupiter: 31557600 * 11.862615,
    saturn: 31557600 * 29.447498,
    uranus: 31557600 * 84.016846,
    neptune: 31557600 * 164.79132,
  }

  def initialize(age)
    @age = age.to_f
  end

  AGE_CONVERSION.each do |planet, convert_to_planet_age|
    define_method("on_#{planet}") do
      @age / convert_to_planet_age
    end
  end
end

File: cpp/space-age/space_age.cpp
#include "space_age.h"

namespace space_age
{
    space_age::space_age(const unsigned long long &seconds)
    {
        _seconds = seconds;
    }

    double space_age::seconds() const
    {
        return _seconds;
    }

    double space_age::on_mercury() const
    {
        return _seconds / PLANETS_CONVERTION.at("mercury");
    }

    double space_age::on_venus() const
    {
        return _seconds / PLANETS_CONVERTION.at("venus");
    }

    double space_age::on_earth() const
    {
        return _seconds / PLANETS_CONVERTION.at("earth");
    }

    double space_age::on_mars() const
    {
        return _seconds / PLANETS_CONVERTION.at("mars");
    }

    double space_age::on_jupiter() const
    {
        return _seconds / PLANETS_CONVERTION.at("jupiter");
    }

    double space_age::on_saturn() const
    {
        return _seconds / PLANETS_CONVERTION.at("saturn");
    }

    double space_age::on_uranus() const
    {
        return _seconds / PLANETS_CONVERTION.at("uranus");
    }

    double space_age::on_neptune() const
    {
        return _seconds / PLANETS_CONVERTION.at("neptune");
    }
} // namespace space_age
